core/src/main/scala/kafka/admin/ConsumerGroupCommand.scala

```diff
@@ -165,12 +124,141 @@ object ConsumerGroupCommand extends Logging {
                                                 clientId: Option[String], logEndOffset: Option[Long])
 
+  protected case class MemberAssignmentState(group: String, consumerId: Option[String], host: Option[String],
+                                             clientId: Option[String], numPartitions: Option[Int],
+                                             assignment: Option[List[TopicPartition]])
+
+  protected case class GroupState(group: String, assignmentStrategy: Option[String], state: Option[String], numMembers: Option[Int])
+
   sealed trait ConsumerGroupService {
 
     def listGroups(): List[String]
 
-    def describeGroup(): (Option[String], Option[Seq[PartitionAssignmentState]]) = {
-      collectGroupAssignment(opts.options.valueOf(opts.groupOpt))
+    private def checkGroup(group: String, state: Option[String], numRows: Option[Int]): Boolean = {
+      var printList = true
+      numRows.isDefined match {
+        case false =>
+          // applies to both old and new consumer
+          printError(s"The consumer group '$group' does not exist.")
+          printList = false
+        case true =>
+          if (!opts.useOldConsumer) {
+            state match {
+              case Some("Dead") =>
+                printError(s"Consumer group '$group' does not exist.")
+                printList = false
+              case Some("Empty") =>
+                Console.err.println(s"Consumer group '$group' has no active members.")
+                if (numRows.get == 0)
+                  printList = false
+              case Some("PreparingRebalance") | Some("CompletingRebalance") =>
+                Console.err.println(s"Warning: Consumer group '$group' is rebalancing.")
+              case Some("Stable") =>
+              case other =>
+                // the control should never reach here
+                throw new KafkaException(s"Expected a valid consumer group state, but found '${other.getOrElse("NONE")}'.")
+            }
+          }
+      }
+
+      printList
+    }
+
+    private def size(colOpt: Option[Seq[Object]]): Option[Int] =
+      colOpt.flatMap(col => Try(col.size).toOption)
+
+    private def printOffsets(group: String, state: Option[String], assignments: Option[Seq[PartitionAssignmentState]]): Unit = {
+      if (checkGroup(group, state, size(assignments))) {
+        // find proper columns width
+        var (maxTopicLen, maxConsumerIdLen, maxHostLen) = (15, 15, 15)
+        assignments.get.foreach { consumerAssignment =>
+          maxTopicLen = Math.max(maxTopicLen, consumerAssignment.topic.getOrElse(MISSING_COLUMN_VALUE).length)
+          maxConsumerIdLen = Math.max(maxConsumerIdLen, consumerAssignment.consumerId.getOrElse(MISSING_COLUMN_VALUE).length)
+          if (!opts.useOldConsumer)
+            maxHostLen = Math.max(maxHostLen, consumerAssignment.host.getOrElse(MISSING_COLUMN_VALUE).length)
+        }
+
+        print(s"\n%${-maxTopicLen}s %-10s %-15s %-15s %-15s %${-maxConsumerIdLen}s "
```

Comment:
Not sure if there's a nicer way to do this. I know it's also possible to include the formatter behind the substitution. For example: `$foo%-15s` would apply the formatter to whatever is substituted by `foo`.

Reply:
I was referring to the mix of string interpolation and format, which is a bit odd. It seems like interpolation already provides the formatting facilities you're using here.